### PR TITLE
Add a Publication Search Service for background search

### PR DESCRIPTION
**Pull request by mickael-menu**
### Added

* A new Publication `SearchService` to search through the resources' content, with a default implementation `StringSearchService`.

### CODE REVIEW

File: r2-shared/src/main/java/org/readium/r2/shared/publication/services/search/StringSearchService.kt [hunk 1]
```diff
@@ -80,6 +80,10 @@ class StringSearchService(
         }
 
     private inner class Iterator(val publication: Publication, val query: String, val options: Options, val locale: Locale) : SearchIterator {
+
+        override var resultCount: Int? = 0
```
Comment (mrifni) [inline r2-shared/src/main/java/org/readium/r2/shared/publication/services/search/StringSearchService.kt:84]: is it possible to have `lateinit` or not nullable type here if 0 is assigned by default ?

File: r2-shared/src/main/java/org/readium/r2/shared/publication/services/search/StringSearchService.kt [hunk 2]
```diff
@@ -103,6 +107,8 @@ class StringSearchService(
                 }
 
                 val locators = findLocators(index, link, text)
+                resultCount = (resultCount ?: 0) + locators.count()
```
Comment (mrifni) [inline r2-shared/src/main/java/org/readium/r2/shared/publication/services/search/StringSearchService.kt:110]:
```suggestion
                resultCount = resultCount + locators.count()
```

Might not need null check at all if type is converted to not nullable

File: r2-shared/src/main/java/org/readium/r2/shared/publication/services/search/StringSearchService.kt (hunk 1 above)
Comment (mickael-menu) [reply r2-shared/src/main/java/org/readium/r2/shared/publication/services/search/StringSearchService.kt:84]: `resultCount` is an overridden property defined as `Int?` in the `SeachIterator` interface, but in this particular implementation it always have a value.

File: r2-shared/src/main/java/org/readium/r2/shared/publication/services/search/StringSearchService.kt (hunk 2 above)
Comment (mickael-menu) [reply r2-shared/src/main/java/org/readium/r2/shared/publication/services/search/StringSearchService.kt:110]: See https://github.com/readium/r2-shared-kotlin/pull/154#discussion_r635273765

File: r2-shared/src/main/java/org/readium/r2/shared/publication/services/search/StringSearchService.kt (hunk 1 above)
Comment (mickael-menu) [reply r2-shared/src/main/java/org/readium/r2/shared/publication/services/search/StringSearchService.kt:84]: Actually it looks like we can override a nullable type into a non-nullable one, so your suggestion is definitely an improvement. Thanks!

**Comment by mrifni** [comment]
I am eagerly waiting on this change, whats like the time frame to get this PR merged and what part of the release this will go into ?

**Comment by mickael-menu** [comment]
@mrifni For a release we'll have to wait a few more weeks, there's also the new audiobook navigator to merge in before. But we could merge this in `develop` pretty quickly. Would you be able to test from `develop`?

**Comment by mickael-menu** [comment]
Btw not to get your hopes up too high, the current implementation finds search results and opens the publication at the right position. But the search results are not highlighted in the navigator. This will need a separate "decorator" feature which will come later. This is my next subject.

**Comment by mrifni** [comment]
sure, I can test it. Highlight fix is my next big expectations :)

**Pull request merged**